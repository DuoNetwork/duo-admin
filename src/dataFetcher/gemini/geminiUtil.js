'use strict';
const MysqlUtil = require('../../utils/mysqlUtil');
import * as CST from '../../constant';

let dbConn;

const EXCHANGE_NAME = CST.EXCHANGE_GEMINI;
const DB_HOST = CST.DB_HOST;
const DB_USER = CST.DB_USER;
const DB_PASSWORD = CST.DB_PASSWORD;
const DB_PRICEFEED = CST.DB_PRICEFEED;
const DB_TABLE_TRADE = CST.DB_TABLE_TRADE;

class GeminiTradeFeedUtil {
	initDB() {
		console.log('Init the DB');

		mysqlUtil.setup(EXCHANGE_NAME, DB_HOST, DB_USER, DB_PASSWORD, DB_PRICEFEED, DB_TABLE_TRADE);

		mysqlUtil.initDB();
	}

	fetchETHTradesByOwnWebSocket() {
		const ws = require('ws');
		const w = new ws('wss://api.gemini.com/v1/marketdata/ETHUSD');

		w.on('message', msg => {
			var parsedJson = JSON.parse(msg);

			if (parsedJson.events[0].type == 'trade') {
				console.log(parsedJson.events[0]);

				var item = parsedJson.events[0];

				dbConn = mysqlUtil.dbConn;
				if (dbConn == undefined) {
					geminiTradeFeedUtil.initDB();
				}

				var trade_type = 'buy';

				if (item.makerSide == 'ask') {
					trade_type = 'buy';
				} else if (item.makerSide == 'bid') {
					trade_type = 'sell';
				}

				// no timestamp returned by exchange so we leave empty there.
				mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, item.tid, item.price, item.amount, trade_type, '');
			}
		});

		w.on('open', () => {
			console.log('[Gemini]-WebSocket is open');
		});
	}
}
let mysqlUtil = new MysqlUtil();
let geminiTradeFeedUtil = new GeminiTradeFeedUtil();
export default geminiTradeFeedUtil;

/*

{ type: 'trade',
  tid: 3553004418,
  price: '626.99',
  amount: '0.15791304',
  makerSide: 'ask' }

  */
