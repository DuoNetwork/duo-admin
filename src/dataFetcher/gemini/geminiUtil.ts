import MysqlUtil from '../../utils/MysqlUtil';
import * as CST from '../../constant';

const EXCHANGE_NAME = CST.EXCHANGE_GEMINI;
const DB_HOST = CST.DB_HOST;
const DB_USER = CST.DB_USER;
const DB_PASSWORD = CST.DB_PASSWORD;
const DB_PRICEFEED = CST.DB_PRICEFEED;
const DB_TABLE_TRADE = CST.DB_TABLE_TRADE;

export class GeminiTradeFeedUtil {
	mysqlUtil: MysqlUtil;

	constructor() {
		console.log('begin');
		this.mysqlUtil = new MysqlUtil(EXCHANGE_NAME, DB_HOST, DB_USER, DB_PASSWORD, DB_PRICEFEED, DB_TABLE_TRADE);
	}

	initDB() {
		console.log('Init the DB');
		this.mysqlUtil.initDB();
	}

	fetchETHTradesByOwnWebSocket() {
		const ws = require('ws');
		const w = new ws('wss://api.gemini.com/v1/marketdata/ETHUSD');

		w.on('message', msg => {
			const parsedJson: any = JSON.parse(msg);

			if (parsedJson.events[0].type == 'trade') {
				var timestampms = parsedJson.timestampms;

				// console.log(parsedJson.events[0]);

				const item = parsedJson.events[0];

				const dbConn = this.mysqlUtil.dbConn;
				if (dbConn == undefined) {
					geminiTradeFeedUtil.initDB();
				}

				let trade_type = 'buy';

				if (item.makerSide == 'ask') {
					trade_type = 'buy';
				} else if (item.makerSide == 'bid') {
					trade_type = 'sell';
				}

				if (timestampms == undefined) {
					timestampms = '';
				}

				// no timestamp returned by exchange so we leave empty there.
				this.mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, item.tid, item.price, item.amount, trade_type, timestampms);
			}
		});

		w.on('open', () => {
			console.log('[Gemini]-WebSocket is open');
		});
	}
}

const geminiTradeFeedUtil = new GeminiTradeFeedUtil();
export default geminiTradeFeedUtil;
