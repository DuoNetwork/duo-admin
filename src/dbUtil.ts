import dynamoUtil from './database/dynamoUtil';
import sqlUtil from './database/sqlUtil';
import keyUtil from './keyUtil';
import { IOption, IPrice, ITrade } from './types';
import util from './util';
const localSQLauth = require('./keys/mysql.json');

class DbUtil {
	private dynamo: boolean = false;

	public async init(tool: string, option: IOption) {
		this.dynamo = option.dynamo;
		const role = util.getDynamoRole(tool, option.dynamo);
		const process = util.getStatusProcess(tool, option);
		util.log('role: ' + role);
		util.log('process: ' + process);

		dynamoUtil.init(option.live, role, process);
		if ((['bitfinex', 'gemini', 'kraken', 'gdax', 'commit'].includes(tool) && !option.dynamo) || option.server) {
			const sqlAuth = await keyUtil.getSqlAuth(option);
			sqlUtil.init(sqlAuth.host, sqlAuth.user, sqlAuth.password);
		} else
			sqlUtil.init(localSQLauth.host, localSQLauth.user, localSQLauth.password);
	}

	public insertTradeData(trade: ITrade, insertStatus: boolean) {
		return this.dynamo
			? dynamoUtil.insertTradeData(trade, insertStatus)
			: sqlUtil.insertTradeData(trade, insertStatus);
	}

	public insertPrice(price: IPrice) {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.insertPrice(price);
	}

	public readLastPrice(): Promise<IPrice> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readLastPrice();
	}

	public readSourceData(currentTimestamp: number): Promise<ITrade[]> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readSourceData(currentTimestamp);
	}

	public cleanDB(): Promise<void> {
		return this.dynamo ? Promise.reject() : sqlUtil.cleanDB();
	}

	public insertHeartbeat(data: object = {}): Promise<void> {
		return dynamoUtil.insertHeartbeat(data);
	}
}

const dbUtil = new DbUtil();
export default dbUtil;
