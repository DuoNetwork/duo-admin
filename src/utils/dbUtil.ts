import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from '../common/constants';
import { IOption, IPriceFix, ITrade } from '../common/types';
import dynamoUtil from './dynamoUtil';
import keyUtil from './keyUtil';
import sqlUtil from './sqlUtil';
import util from './util';

class DbUtil {
	public dynamo: boolean = false;

	public async init(tool: string, option: IOption, web3Wrapper: Web3Wrapper) {
		this.dynamo = option.dynamo;
		const process = util.getStatusProcess(tool, option);
		util.logInfo('process: ' + process);

		let config = {};
		try {
			config = require('../keys/aws/' + (option.live ? 'live' : 'dev') + '/admin.json');
		} catch (error) {
			util.logError(error);
		}
		await dynamoUtil.init(
			config,
			option.live,
			process,
			(value: string | number) => web3Wrapper.fromWei(value),
			async txHash => {
				const txReceipt = await web3Wrapper.getTransactionReceipt(txHash);
				if (!txReceipt) return null;
				return {
					status: txReceipt.status
				};
			}
		);
		if ([CST.TRADES, CST.COMMIT, CST.CLEAN_DB].includes(tool) && !option.dynamo) {
			const sqlAuth = await keyUtil.getSqlAuth(option);
			return sqlUtil.init(sqlAuth.host, sqlAuth.user, sqlAuth.password);
		}
	}

	public insertTradeData(trade: ITrade, insertStatus: boolean) {
		return this.dynamo
			? dynamoUtil.insertTradeData(trade, insertStatus)
			: sqlUtil.insertTradeData(trade, insertStatus);
	}

	public insertPrice(price: IPriceFix) {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.insertPrice(price);
	}

	public readLastPrice(quote: string, base: string): Promise<IPriceFix> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readLastPrice(quote, base);
	}

	public readSourceData(
		currentTimestamp: number,
		quote: string,
		base: string
	): Promise<ITrade[]> {
		return this.dynamo
			? Promise.reject('invalid')
			: sqlUtil.readSourceData(currentTimestamp, base, quote);
	}

	public cleanDB(): Promise<void> {
		return this.dynamo ? Promise.reject() : sqlUtil.cleanDB();
	}

	public insertHeartbeat(data: object = {}): Promise<void> {
		return dynamoUtil.insertHeartbeat(data);
	}
}

const dbUtil = new DbUtil();
export default dbUtil;
