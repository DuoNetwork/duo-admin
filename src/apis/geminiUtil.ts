import ws from 'ws';
import * as CST from '../constants';
import dbUtil from '../dbUtil';
import { ITrade } from '../types';
import util from '../util';

export class GeminiUtil {
	public parseTrade(parsedJson: any): ITrade {
		const timestampms = parsedJson.timestampms || '';
		const trade = parsedJson.events[0];

		return {
			source: CST.EXCHANGE_GEMINI,
			id: trade.tid + '',
			price: Number(trade.price),
			amount: Math.abs(Number(trade.amount)),
			timestamp: Number(timestampms)
		};
	}

	public parseApiResponse(msg: string) {
		const parsedJson: any = JSON.parse(msg);
		// util.log(parsedJson);

		if (parsedJson.events[0].type === 'trade') {
			// util.log(parsedJson);
			const parsedTrade: ITrade = this.parseTrade(parsedJson);

			// no timestamp returned by exchange so we leave empty there.
			dbUtil.insertSourceData(parsedTrade);
			util.log(CST.EXCHANGE_GEMINI + ': trade fetched and inserted ' + parsedTrade.id);
		}
	}

	public fetchTradesWithoutRetry() {
		const w = new ws('wss://api.gemini.com/v1/marketdata/ETHUSD');

		w.on('message', msg => this.parseApiResponse(msg.toString()));

		w.on('open', () => util.log(CST.EXCHANGE_GEMINI + ': webSocket is open'));

		w.on('close', (code: number, reason: string) => {
			util.log(CST.EXCHANGE_GEMINI + ': ' + code + ' ' + reason);
			throw new Error('ws closed');
		});

		w.on('error', (error: Error) => util.log(CST.EXCHANGE_GEMINI + ': ' + error));
	}

	public fetchTrades() {
		try {
			this.fetchTradesWithoutRetry();
		} catch (err) {
			util.log(CST.EXCHANGE_GEMINI + ': ' + err);
			this.fetchTrades();
		}
	}
}

const geminiUtil = new GeminiUtil();
export default geminiUtil;
