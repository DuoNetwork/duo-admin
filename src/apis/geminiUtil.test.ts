import dbUtil from '../dbUtil';
import geminiUtil from './geminiUtil';
const trades: Array<{ [key: string]: string }> = require('../samples/gemini.json');
const apiResponse = JSON.stringify(require('../samples/geminiMsg.json'));

test('parseTrade', () => {
	trades.forEach(async trade => {
		const parsedTrade = geminiUtil.parseTrade(trade);
		expect(parsedTrade).toMatchSnapshot();
	});
});

test('parseApiResponse', async () => {
	dbUtil.insertSourceData = jest.fn(() => Promise.resolve());
	await geminiUtil.parseApiResponse(apiResponse);
	// for (let i = 0; i < 6; i++)
	expect(
		(dbUtil.insertSourceData as jest.Mock<Promise<void>>).mock.calls[0][0]
	).toMatchSnapshot();
});
