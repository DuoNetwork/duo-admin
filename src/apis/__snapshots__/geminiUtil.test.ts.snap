// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseApiResponse 1`] = `
Object {
  "amount": "50",
  "price": "654.00",
  "source": "GEMINI",
  "sourceTimestamp": 1524810712835,
  "tradeId": 3590632229,
  "tradeType": "buy",
}
`;

exports[`parseTrade 1`] = `
Object {
  "amount": "0.027783",
  "price": "629.69",
  "source": "GEMINI",
  "sourceTimestamp": 1524739084253,
  "tradeId": 3584809531,
  "tradeType": "buy",
}
`;

exports[`parseTrade 2`] = `
Object {
  "amount": "0.021431",
  "price": "629.79",
  "source": "GEMINI",
  "sourceTimestamp": 1524739191718,
  "tradeId": 3584814698,
  "tradeType": "buy",
}
`;
